spec/formatting.md

```diff
@@ -18,3 +18,38 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Bidirectional Isolation
+
+Formatted messages may be bidirectional,
+i.e. contain a mix of left-to-right and right-to-left content.
+In all output formats,
+such parts should be isolated from each other so that
+the directionality of one part of the message does not affect
+the presentation of the rest of the message.
+
+Implementations may provide one or more strategies for isolation,
+but one such strategy must behave as follows,
+when formatting the message as a string:
+
+Considering a "formatted part" to be the formatted string representation of a Placeholder,
+when the formatted part has a direction different from the message as a whole
+or when its directionality is not known,
+the formatted string representation of the message must
+prefix the formatted part with one of the following isolate characters:
+
+- U+2066 LEFT-TO-RIGHT ISOLATE if the formatted part is known to have LTR directionality,
+- U+2067 RIGHT-TO-LEFT ISOLATE if the formatted part is known to have RTL directionality, or
+- U+2068 FIRST STRONG ISOLATE if the formatted part's directionality is not known.
+
+In all cases where a formatted part is prefixed with such an isolate character,
+it must be postfixed with a corresponding U+2069 POP DIRECTIONAL ISOLATE character.
+
+An implementation must define its own method for
+determining the directionality of a formatted part,
+relying either on metadata attached to the formatted part,
+or on introspection of its contents.
+
+For example, an implementation could provide a `:number` formatting function
+which would always produce output matching the message's locale,
+allowing for its formatted string representation to never need isolation.
```

Comment:
You're assuming that the message formatter's locale matches the language of the pattern, which is a bad assumption. Many (most!) resource systems fall back or allow sparse population of translated resources. As a result, the locale of the message formatter (and its subformatters) may not match that of the pattern string. _This is a feature._ It is often the case that developers will wish to test code in a given locale before the translations are available or to use pseudo.

I also note that this paragraph is not an example of the preceding requirement. Perhaps insert this just ahead of this paragraph:

> Implementations _should not_ isolate formatted parts when the base paragraph direction of the pattern and the formatted parts match *and* there are no opposite direction runs of text in the formatted part.

Reply:
Let me give an example to verify if I've understood your concern right:

An app using English as its fallback locale is partially localized to Arabic, with some English messages filling in holes. If such a LTR message is formatted using an RTL formatter _and_ it contains a placeholder with a variable reference to an RTL value, we should isolate the placeholder's value in the output, even though the formatter and the value are both RTL.

If this is the case, then I think what we need is a more explicit definition somewhere of the "message locale", which I would consider to be English in the above example, even if the formatter is nominally in Arabic.

Reply:
That's more or less effectively what I'm saying. The challenge is that (let's use Java as an example), I have code that looks like this:

```
ResourceBundle myPatterns = ResourceBundle.getBundle("com.example.Patterns", myRuntimeLocale);
MessageFormat2 formatter = new MessageFormat2(myPatterns.getString("somePattern"), myRuntimeLocale);
```

If you query `myPatterns` for its locale, you get `myRuntimeLocale`, even though the bundle is really a stack of bundles. If `myRuntimeLocale` is `ar-AE`, the actual translations probably live in the `ar` bundle. If the developer has added a new string, though, `somePattern` may be coming from the `ROOT` locale bundle (and be in, e.g. English). Both `formatter` and its subformat patterns are trying to serve the `ar-AE` locale, though and finding out which specific locale bundle coughed up the pattern is both inaccessible to it (MF2 doesn't know that the pattern is coming from a bundle--it's just a string, as far as it is concerned) and doesn't add any value to instrument in most cases.

The actual language (as opposed to the asserted language/locale of it) is not really a concern for MF2, as long as the pattern is valid. If we wanted to do language-sensitive processing (capitalization, for example), we'd still use a locale for that. 

When it comes to bidi, it also doesn't matter as much as the base paragraph direction does, save that the language might serve as a hint for what the direction should be (lacking other information).